builtin/providers/azurerm/resource_arm_container_service.go

```diff
@@ -0,0 +1,669 @@
+		agentPoolProfile["count"] = int(*profile.Count)
+		agentPoolProfile["dns_prefix"] = *profile.DNSPrefix
+		agentPoolProfile["fqdn"] = *profile.Fqdn
+		agentPoolProfile["name"] = *profile.Name
+		agentPoolProfile["vm_size"] = string(profile.VMSize)
+		agentPoolProfiles.Add(agentPoolProfile)
+	}
+
+	return &agentPoolProfiles
+}
+
+func flattenAzureRmContainerServiceServicePrincipalProfile(profile *containerservice.ServicePrincipalProfile) *schema.Set {
+
+	if profile == nil {
+		return nil
+	}
+
+	servicePrincipalProfiles := schema.Set{
+		F: resourceAzureRMContainerServiceServicePrincipalProfileHash,
+	}
+
+	values := map[string]interface{}{}
+
+	values["client_id"] = *profile.ClientID
+	if profile.Secret != nil {
+		values["client_secret"] = *profile.Secret
+	}
+
+	servicePrincipalProfiles.Add(values)
+
+	return &servicePrincipalProfiles
+}
+
+func flattenAzureRmContainerServiceDiagnosticsProfile(profile *containerservice.DiagnosticsProfile) *schema.Set {
+	diagnosticProfiles := schema.Set{
+		F: resourceAzureRMContainerServiceDiagnosticProfilesHash,
+	}
+
+	values := map[string]interface{}{}
+
+	values["enabled"] = *profile.VMDiagnostics.Enabled
+	if profile.VMDiagnostics.StorageURI != nil {
+		values["storage_uri"] = *profile.VMDiagnostics.StorageURI
+	}
+	diagnosticProfiles.Add(values)
+
+	return &diagnosticProfiles
+}
+
+func expandAzureRmContainerServiceDiagnostics(d *schema.ResourceData) containerservice.DiagnosticsProfile {
+	configs := d.Get("diagnostics_profile").(*schema.Set).List()
+	profile := containerservice.DiagnosticsProfile{}
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		enabled := data["enabled"].(bool)
+
+		profile = containerservice.DiagnosticsProfile{
+			VMDiagnostics: &containerservice.VMDiagnostics{
+				Enabled: &enabled,
+			},
+		}
+	}
+
+	return profile
+}
+
+func expandAzureRmContainerServiceLinuxProfile(d *schema.ResourceData) containerservice.LinuxProfile {
+	profiles := d.Get("linux_profile").(*schema.Set).List()
+	config := profiles[0].(map[string]interface{})
+
+	adminUsername := config["admin_username"].(string)
+	linuxKeys := config["ssh_key"].([]interface{})
+	sshPublicKeys := []containerservice.SSHPublicKey{}
+
+	for _, key := range linuxKeys {
+
+		sshKey, ok := key.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		keyData := sshKey["key_data"].(string)
+
+		sshPublicKey := containerservice.SSHPublicKey{
+			KeyData: &keyData,
+		}
+
+		sshPublicKeys = append(sshPublicKeys, sshPublicKey)
+	}
+
+	profile := containerservice.LinuxProfile{
+		AdminUsername: &adminUsername,
+		SSH: &containerservice.SSHConfiguration{
+			PublicKeys: &sshPublicKeys,
+		},
+	}
+
+	return profile
+}
+
+func expandAzureRmContainerServiceMasterProfile(d *schema.ResourceData) containerservice.MasterProfile {
+	configs := d.Get("master_profile").(*schema.Set).List()
+	profile := containerservice.MasterProfile{}
+
+	for _, configRaw := range configs {
```

Comment:
This should just use the first item in the list rather than looping, and `fqdn` should be read.